azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -210,38 +217,118 @@ private void createLogger() {
+    // If present, remove the existing file appender
+    if (jobAppender != null) {
+      removeAppender(jobAppender);
+      flowLogger.info("Removed old file appender for job " + this.jobId);
+    }
+
+    jobAppender = appender;
+    logger.addAppender(jobAppender);
+    logger.setAdditivity(false);
+    flowLogger.info("Attached file appender for job " + this.jobId);
+  }
+
+  private FileAppender createFileAppender() throws IOException {
+    // Set up log files
+    String logName = createLogFileName(node);
+    logFile = new File(workingDir, logName);
+    String absolutePath = logFile.getAbsolutePath();
+
+    // Attempt to create FileAppender
+    RollingFileAppender fileAppender =
+        new RollingFileAppender(loggerLayout, absolutePath, true);
+    fileAppender.setMaxBackupIndex(jobLogBackupIndex);
+    fileAppender.setMaxFileSize(jobLogChunkSize);
+
+    flowLogger.info("Created file appender for job " + this.jobId);
+    return fileAppender;
+  }
+
   private void createAttachmentFile() {
     String fileName = createAttachmentFileName(node);
     File file = new File(workingDir, fileName);
     attachmentFileName = file.getAbsolutePath();
   }
 
+  private void attachKafkaAppender(KafkaLog4jAppender appender) {
+    // If present, remove the existing kafka appender
+    if (kafkaAppender.isPresent()) {
+      removeAppender(kafkaAppender);
+      flowLogger.info("Removed old Kafka appender for job " + this.jobId);
+    }
+
+    kafkaAppender = Optional.of(appender);
+    logger.addAppender(kafkaAppender.get());
+    logger.setAdditivity(false);
+    flowLogger.info("Attached new Kafka appender for job " + this.jobId);
+  }
+
+  private KafkaLog4jAppender createKafkaAppender() {
+    KafkaLog4jAppender kafkaProducer = new KafkaLog4jAppender();
+    Props azkProps = AzkabanExecutorServer.getApp().getAzkabanProps();
+    kafkaProducer.setSyncSend(false);
+    kafkaProducer.setBrokerList(azkProps.getString("kafkaLogging.brokerList"));
+    kafkaProducer.setTopic(azkProps.getString("kafkaLogging.topic"));
+
+    JSONObject layout = new JSONObject();
+    layout.put("category", "%c{1}");
+    layout.put("level", "%p");
+    layout.put("message", "%m");
+    layout.put("projectname", props.getString("azkaban.flow.projectname"));
+    layout.put("flowid", props.getString("azkaban.flow.flowid"));
+    layout.put("submituser", props.getString("azkaban.flow.submituser"));
+    layout.put("execid", props.getString("azkaban.flow.execid"));
+    layout.put("projectversion", props.getString("azkaban.flow.projectversion"));
+    layout.put("logsource", "userJob");
+
+    kafkaProducer.setLayout(new EnhancedPatternLayout(layout.toString()));
+    kafkaProducer.activateOptions();
+
+    flowLogger.info("Created kafka appender for " + this.jobId);
+    return kafkaProducer;
+  }
+
+  private void removeAppender(Optional<Appender> appender) {
+    removeAppender(appender.get());
+    appender = Optional.empty();
```

Comment:
Is this step necessary? I thought it would only affect the reference in this method's scope.